File: JavaScript/FindTheDuplicateNumber/FindTheDuplicateNumber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
// (106ms)
// binary search method
var findDuplicate = function(nums) {
    if(!nums || nums.length < 2) {
        return -1;
    }
    var left = 0, right = nums.length;
    while(left < right) {
        let mid = Math.floor((left + right) / 2), count = 0;
        nums.forEach(function(item) {
            if(item <= mid)
                count++;
        });
        if(count > mid) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }
    return left;
};

// (89ms)
// linked list cycle method
var findDuplicate2 = function(nums) {
    if(!nums || nums.length < 2) {
        return -1;
    }
    var slow = nums[0];
    var fast = nums[nums[0]];
    while(slow != fast) {
        slow = nums[slow];
        fast = nums[nums[fast]];
    }
    fast = 0;
    while(slow != fast) {
        slow = nums[slow];
        fast = nums[fast];
    }
    return slow;
};

File: Python/FindTheDuplicateNumber/FindTheDuplicateNumber.py
# coding:utf-8
class Solution(object):
    # (52ms)
    # ref: http://keithschwarz.com/interesting/code/?dir=find-duplicate
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums or len(nums) < 2:
            return -1
        slow, fast = nums[0], nums[nums[0]]
        while slow != fast:
            slow = nums[slow]
            fast = nums[nums[fast]]

        fast = 0
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]
        return slow

    # (65ms)
    # binary search method
    def findDuplicate2(self, nums):
        if not nums or len(nums) < 2:
            return -1
        left, right = 0, len(nums)
        while left < right:
            mid, count = (left + right) / 2, 0
            for num in nums:
                if num <= mid:
                    count += 1
            if count > mid:
                right = mid
            else:
                left = mid + 1
        return left
